### DISCUSSION TITLE
Does carapace provide argument validation for cobra CLIs?

### DISCUSSION THREAD

**Discussion by plastikfan** [carapace-sh/carapace-bin#2599] [General]
I have just discovered this project and am very impressed although I'm still struggling to see how I would integrate with my own cli projects. Initially I thought you just installed caraspace and use it to provide completions for cli programs interactively. But reading a bit more and I discover there is a full blown api too.

This is intriguing to me because I have already written my own cobra assistance library that fills a gap left by cobra in that it doesnt provide a framework for validating arguments after they have been parsed. From what I understand, carapsace works within the pre-parse phase, where as my validation schem works post prase phase so I'm thinking I can use them together, but I am not sure. The carapsace documentation is very good, I havent done a detailed read yet so my current misaprehensions will be resolved after having done so.

But the core of my query is this; does carapsace provide validation after parsing or not? In fact I would be interested in the possibility of integrating my validation scheme directly into caraspace if there would be an opportunity to do so (I'll probably start off with a fork for a feasibility study). I have realised that some parts of what I have implemented (which I did at a very early stage of learning Golang) are either surplace to requirements (as they could be acheived with carapace), or based on a flawed premise (eg my fake enum implementations which are obseleted by completions).

**Comment by rsteube** [comment]
Depends on what you want to do.
You should be able to hook into the post-parse phase with [PreInvoke](https://carapace-sh.github.io/carapace/carapace/gen/preInvoke.html).
If it's about custom flag values, pflag has [Var](https://pkg.go.dev/github.com/spf13/pflag#Var) and [VarP](https://pkg.go.dev/github.com/spf13/pflag#Var).

**Comment by plastikfan** [comment]
Hi @rsteube, my validation scheme already builds upon pflag's Var and VarP, but your PreInvoke looks interetsing. The type of validations that my library provides are these:

- ensure a value is within a range
- ensure a value greater than a min
- ensure a value less than max
- ensure a value matches a regex pattern
- ensure a  string value is one of a subset of pre-defined values (I created a custom enum type to achieve this, but this is now made redundant by the facility to achive this by carapace completions)
- plus others

Since Var/PVar are type specific, the resultant code is repetitve differing only by type, and since Var/PVar is pre-implemented not using generics, I am restricted either by coding the validators up explicitly, or by using code generators. I have opted for the latter, but I wonder if I could replace these to integrate with carapace somehow. I think Actions will do the trick but not sure. What I would want is to prevent a user from entering an invalid value by being able to validate an option value using a pre-defined validator function. How can this be achieved with an Action?

Does the PreInvoke allow for a mechansim where what is typed in by the user is rejected as a result of failing a validation check, eg a numeric field is outside of a predefined range and clears the value so they can correct it, before being accepted? This is the functionality being provided by my validation scheme except it was not based upon completeion and required the user to hit enter, get processed then rejected and the user prompted to start over; clearly this is clumsy compared to the completions approach.

I'm very impressed with the work you have completed with this repository as my knowledge on bulding CLIs with advanced features is greatly increased simply by discovering this repo. This does also mean I have to really reconsider the work I have done on my libarary and wonder if its still valid, perhaps I will trash it.


Mmmm, just looking at the Group Therapy examples showing multiple concurrent completions taking place (presumebaly in separate go routines, is that correct?). The mind boggles with the potential for what this could be used for!

**Reply by rsteube** [reply]
Cobra at least has a validation function for positional args: https://github.com/spf13/cobra/blob/main/args.go

It is possible to do argument validation in `PreInvoke` by returning `ActionMessage` as error.
But using `Value` in pflag and `Args` in in cobra should be preferred.
As this ensures consistent behaviour during completion and running of the command.

More complex validation might be better left to the run function and left out during completion (at least if it is slow).

Seems there are similar attempts using viper (which might be a better angle to hook in): https://stackoverflow.com/a/72138562 .
I think it would be hard for something like this to be adopted unless it is officially provided by cobra.

**Reply by plastikfan** [reply]
I wish I understood completions before I started my project, I would probably not have done it this way if I really understood it. I am already aware of the validation for positionals but my scheme was aimed at option values as these were not covered by cobra. My project is largely invalidated by completions but going forward I'm going to explore Actions more and see what I can achieve with them.

Can I suggest an improvement in your documentation. I don't think you mention anything relating to the terminal. My current terminal is Warp which provides its own ai based completions. It does not allow the shell to do any completion handling whatsoever and this means all the bells and whistles that a developer can provide using carapace is all for nothing because they are not called upon. Because of this single issue, I am in process of trying out different terminals that allow the shell to have the freedom to handle completions; I'm trying Wave at the moment but there are other options. I think it would be useful to cover this in the documentation as I was quite confused about this until I discovered the problem myself.

**Reply by rsteube** [reply]
Warp is pretty much the worst in that regard (and the only one causing an issue like this).

There are some good ones like [foot](https://github.com/DanteAlighierin/foot) (linux+wayland), and [wezterm](https://wezfurlong.org/wezterm/) (linux/mac/win).
But [ghostty](https://mitchellh.com/writing/ghostty-is-coming) is about to be released and will likely gain a large adoption (https://www.youtube.com/watch?v=l_qY2p0OH9A).

**Reply by plastikfan** [reply]
Great thanks @rsteube, I have just joined the ghostty discord server and keenly await its release. In the meantime, I am using Wave. I'm beginning to learn a bit more about carapace but I have also discovered some rough edges (like crashing, so I'll probably raise an issue or two 🤭).
